docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,625 @@
+    - [Proposed contract changes](#proposed-contract-changes)
+      - [Changes to the ControlPlane contract](#changes-to-the-controlplane-contract)
+    - [Proposed controller changes](#proposed-controller-changes)
+      - [Changes to the Machine controller](#changes-to-the-machine-controller)
+      - [Changes to the MachinePool controller](#changes-to-the-machinepool-controller)
+      - [Changes to the Cluster topology, MachineSet and MachineDeployment controllers](#changes-to-the-cluster-topology-machineset-and-machinedeployment-controllers)
+      - [Changes to the cluster-autoscaler](#changes-to-the-cluster-autoscaler)
+  - [Security Model](#security-model)
+    - [Migration strategy for Bootstrap providers](#migration-strategy-for-bootstrap-providers)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Alternatives](#alternatives)
+  - [Continuous reconciliation](#continuous-reconciliation)
+  - [Type definition of a taint in Cluster API objects](#type-definition-of-a-taint-in-cluster-api-objects-1)
+- [Upgrade Strategy](#upgrade-strategy)
+- [Additional Details](#additional-details)
+  - [Kubernetes taints API discussions](#kubernetes-taints-api-discussions)
+  - [Test Plan [optional]](#test-plan-optional)
+  - [Graduation Criteria [optional]](#graduation-criteria-optional)
+- [Implementation History](#implementation-history)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+This proposal introduces taints as a first-class citizen in Cluster API's core types, enabling users to declaratively manage Node taints through higher-level Cluster API resources such as Cluster, ClusterClass, MachineSet, MachineDeployment and MachinePool.
+
+The proposal defines two different kind of propagation modes for taints:
+
+- **Always**: Taints are continuously reconciled and maintained on nodes
+- **OnInitialization**: Taints are set once during node initialization and then left unmanaged
+
+**Note:** This new proposal has been created rather than updating the prior [in-place metadata propagation](20221003-In-place-propagation-of-Kubernetes-objects-only-changes.md) proposal because taints are not yet part of the Core Provider's API types and are different enough from labels or annotations that a different set of constraints will need to be considered.
+Very early versions of Kubernetes tracked taints as annotations, but they have long since been [promoted to their own API type](https://github.com/kubernetes/kubernetes/commit/9b640838a5f5e28db1c1f084afa393fa0b6d1166)
+
+## Motivation
+
+As stated in the goals of the project, Cluster API tries to "define common operations" and "manage the lifecycle [...] of Kubernetes-conformant Clusters using a declarative API"
```

Comment:
nit: I think [...] doesn't work as a link. At least not on GitHub: https://github.com/nrb/cluster-api/blob/taints-propagation-proposal/docs/proposals/20250513-propogate-taints.md

Reply:
Lol. My bad. I thought [...] was also supposed to be a link 😂